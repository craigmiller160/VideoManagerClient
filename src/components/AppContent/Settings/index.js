/*
 *     VideoManagerClient
 *     Copyright (C) 2020 Craig Miller
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import classes from './Settings.module.scss';
import FlexRow from '../../UI/Grid/FlexRow';
import Spinner from '../../UI/Spinner/Spinner';
import {
	loadSettings,
	saveSettings
} from '../../../store/settings/settings.actions';
import Form from '../../UI/form/Form/Form';
import Input from '../../UI/form/Input/Input';
import { Button } from 'reactstrap';
import { isRequired } from '../../../utils/validations';
import FileChooser from '../../UI/FileChooser';
import { change } from 'redux-form';
import ToolTip from '../../UI/ToolTip';

export const FORM_NAME = 'Settings_Form';

const Settings = (props) => {
	const { rootDirEditing, rootDirModified } = props;
	const dispatch = useDispatch();
	const loading = useSelector((state) => state.settings.loading);
	const form = useSelector((state) => state.form[FORM_NAME], shallowEqual);
	const [state, setState] = useState({
		rootDirEditing,
		rootDirModified
	});

	useEffect(() => {
		dispatch(loadSettings());
	}, []);

	const editRootDir = () =>
		setState((prevState) => ({
			...prevState,
			rootDirEditing: true
		}));

	const selectDir = (selected) => {
		setState((prevState) => ({
			...prevState,
			rootDirEditing: false,
			rootDirModified: true
		}));
		dispatch(change(FORM_NAME, 'rootDir', selected.filePath));
	};

	const submit = async (values) => {
		const successful = await dispatch(saveSettings(values));
		if (successful) {
			setState((prevState) => ({
				...prevState,
				rootDirModified: false
			}));
		}
	};

	const showFileChooserClass = state.rootDirEditing ? classes.show : '';
	const showSaveClass = state.rootDirEditing ? '' : classes.show;

	// Separate variable for this to make it easily extensible
	const enableSaveBtn = state.rootDirModified;
	const rootDirToolTipText = form?.values?.rootDir ?? '';

	return (
		<div className={classes.Settings}>
			<FlexRow>
				<div className={classes.title}>
					<h3>Settings</h3>
				</div>
			</FlexRow>
			<Form
				form={FORM_NAME}
				onSubmit={submit}
				destroyOnUnmount={false}
				className={classes.form}
			>
				{loading && <Spinner />}
				{!loading && (
					<div id="settings-form-content">
						<FlexRow
							id="root-dir-container"
							className={classes.rootDirWrapper}
							justifyContent="center"
							alignItems="flex-end"
						>
							<ToolTip text={rootDirToolTipText}>
								<Input
									name="rootDir"
									label="Directory to Scan"
									divClassName={classes.rootDir}
									validate={[isRequired]}
									disabled
								/>
							</ToolTip>
							<Button
								id="set-root-dir-btn"
								color="info"
								onClick={editRootDir}
							>
								Set
							</Button>
						</FlexRow>
						<FlexRow
							id="file-chooser-container"
							justifyContent="center"
							className={[
								classes.fileChooser,
								showFileChooserClass
							].join(' ')}
						>
							<FileChooser
								directoriesOnly
								selectFile={selectDir}
								initialDir={form?.values?.rootDir}
							/>
						</FlexRow>
						<FlexRow
							id="btn-container"
							justifyContent="center"
							className={[classes.submit, showSaveClass].join(
								' '
							)}
						>
							<Button
								id="save-btn"
								type="submit"
								color="primary"
								disabled={!enableSaveBtn}
							>
								Save
							</Button>
						</FlexRow>
					</div>
				)}
			</Form>
		</div>
	);
};
Settings.propTypes = {
	rootDirEditing: PropTypes.bool,
	rootDirModified: PropTypes.bool
};
Settings.defaultProps = {
	rootDirEditing: false,
	rootDirModified: false
};

export default Settings;
