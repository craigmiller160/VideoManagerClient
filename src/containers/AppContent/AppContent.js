import React, { Component } from 'react';
import VideoNavbar from '../../components/VideoNavbar/VideoNavbar';
import Layout from '../../components/Layout/Layout';

class AppContent extends Component {
    render() {
        return (
            <div>
                <VideoNavbar />
                <Layout />
            </div>
        );
    }
}

export default AppContent;