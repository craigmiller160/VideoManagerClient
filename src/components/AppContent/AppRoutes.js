/*
 *     VideoManagerClient
 *     Copyright (C) 2020 Craig Miller
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Redirect, Switch } from 'react-router-dom';
import Scanning from './Scanning/Scanning';
import VideoFileEdit from './VideoFileEdit/VideoFileEdit';
import ManageVideoFilters from './ManageVideoFilters/ManageVideoFilters';
import VideoPlayerPage from './VideoPlayer/VideoPlayerPage';
import VideoListLayout from './VideoListLayout/VideoListLayout';
import ProtectedRoute from '@craigmiller160/react-protected-route';
import Home from './Home/Home';
import { useSelector } from 'react-redux';
import {
    hasAdminRole as hasAdminRoleSelector,
    hasEditRole as hasEditRoleSelector
} from '../../store/auth/auth.selectors';
import Settings from './Settings';

const AppRoutes = (props) => {
    const hasEditRole = useSelector(hasEditRoleSelector);
    const hasAdminRole = useSelector(hasAdminRoleSelector);
    const {
        selectedVideo,
        saveFileChanges,
        isScanning,
        isAuth
    } = props;

    const hasEditRoleRule = {
        allow: ({ editRole }) => editRole,
        redirect: '/'
    };

    const hasAdminRoleRule = {
        allow: ({ adminRole }) => adminRole,
        redirect: '/'
    };

    const isAuthenticatedRule = {
        allow: ({ auth }) => auth,
        redirect: '/login'
    };

    // const isNotAuthenticatedRule = {
    //     allow: ({ auth }) => !auth,
    //     redirect: '/'
    // };

    const isScanningRule = {
        allow: ({ scanning }) => scanning,
        redirect: '/'
    };

    const isNotScanningRule = {
        allow: ({ scanning }) => !scanning,
        redirect: '/scanning'
    };

    const hasSelectedVideoRule = {
        allow: ({ selected }) => !!selected?.fileName,
        redirect: '/'
    };

    const ruleProps = {
        selected: selectedVideo,
        adminRole: hasAdminRole,
        editRole: hasEditRole,
        auth: isAuth,
        scanning: isScanning
    };

    return (
        <Switch>
            <ProtectedRoute
                path="/scanning"
                component={ Scanning }
                ruleProps={ ruleProps }
                rules={ [
                    isAuthenticatedRule,
                    isScanningRule
                ] }
            />
            <ProtectedRoute
                path="/edit"
                component={ VideoFileEdit }
                componentProps={ {
                    selectedVideo,
                    saveFileChanges
                } }
                ruleProps={ ruleProps }
                rules={ [
                    isAuthenticatedRule,
                    isNotScanningRule,
                    hasSelectedVideoRule,
                    hasEditRoleRule
                ] }
            />
            <ProtectedRoute
                path="/filters"
                component={ ManageVideoFilters }
                ruleProps={ ruleProps }
                rules={ [
                    isAuthenticatedRule,
                    isNotScanningRule,
                    hasEditRoleRule
                ] }
            />
            <ProtectedRoute
                path="/play/:fileId"
                component={ VideoPlayerPage }
                ruleProps={ ruleProps }
                rules={ [
                    isAuthenticatedRule,
                    isNotScanningRule
                ] }
            />
            <ProtectedRoute
                path="/videos"
                component={ VideoListLayout }
                ruleProps={ ruleProps }
                rules={ [
                    isAuthenticatedRule,
                    isNotScanningRule
                ] }
            />
            <ProtectedRoute
                component={ Settings }
                ruleProps={ ruleProps }
                path="/settings"
                rules={ [
                    isAuthenticatedRule,
                    hasAdminRoleRule
                ] }
            />
            <ProtectedRoute
                path="/"
                exact
                component={ Home }
                ruleProps={ ruleProps }
            />
            <Redirect to="/" />
        </Switch>
    );
};
AppRoutes.propTypes = {
    selectedVideo: PropTypes.object,
    saveFileChanges: PropTypes.func,
    isScanning: PropTypes.bool,
    isAuth: PropTypes.bool
};

export default AppRoutes;
