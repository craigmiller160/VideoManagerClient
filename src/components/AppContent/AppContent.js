import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import VideoNavbar from 'components/AppContent/VideoNavbar/VideoNavbar';
import { Col, Container, Row } from 'reactstrap';
import Alert from '../UI/Alert/Alert';
import { checkIsScanning, startFileScan } from 'store/scanning/scanning.actions';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import useReactRouter from 'use-react-router';
import { hideAlert, showErrorAlert } from 'store/alert/alert.actions';
import { getSelectedVideoWithFilters } from 'store/videoList/videoList.selectors';
import classes from './AppContent.scss';
import { loadFilterOptions } from 'store/videoSearch/videoSearch.actions';
import { saveVideoFileEdits } from 'store/videoList/videoList.actions';
import AppRoutes from './AppRoutes';
import { checkAuth, logout } from '../../store/auth/auth.actions';

export const AppContent = (props) => {
    const [ isStarted, setStarted ] = useState(false);
    const { history } = useReactRouter();
    const {
        checkIsScanning,
        saveVideoFileEdits,
        isScanning,
        selectedVideo,
        startFileScan,
        alert,
        hideAlert,
        isAuth,
        logout,
        checkAuth,
        loadFilterOptions
    } = props;

    useEffect(() => {
        const doCheckAuth = async () => {
            await checkAuth();
            setStarted(true);
        };
        doCheckAuth();
    }, []);

    useEffect(() => {
        const startup = async () => {
            if (isAuth) {
                loadFilterOptions();
                await checkIsScanning();
            }
        };
        startup();
    }, [isAuth]);

    const saveFileChanges = async () => {
        await saveVideoFileEdits();
        history.push('/');
    };

    return (
        <div className={ classes.AppContent }>
            <VideoNavbar
                startFileScan={ startFileScan }
                disabled={ isScanning || !isAuth }
                logout={ logout }
            />
            {
                isStarted &&
                <Container className={ classes.container }>
                    <Row>
                        <Col xs={{ size: 8, offset: 2 }}>
                            <Alert
                                alert={ alert }
                                hideAlert={ hideAlert }
                            />
                        </Col>
                    </Row>
                    <AppRoutes
                        saveFileChanges={ saveFileChanges }
                        checkIsScanning={ checkIsScanning }
                        selectedVideo={ selectedVideo }
                        isScanning={ isScanning }
                        isAuth={ isAuth }
                    />
                </Container>
            }
        </div>
    );
};

AppContent.propTypes = {
    isScanning: PropTypes.bool,
    alert: PropTypes.object,
    selectedVideo: PropTypes.object,
    checkIsScanning: PropTypes.func,
    startFileScan: PropTypes.func,
    showErrorAlert: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
    loadFilterOptions: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
    hideAlert: PropTypes.func,
    saveVideoFileEdits: PropTypes.func,
    checkAuth: PropTypes.func, // eslint-disable-line react/no-unused-prop-types
    isAuth: PropTypes.bool,
    logout: PropTypes.func
};

const mapStateToProps = (state) => ({
    isScanning: state.scanning.isScanning,
    alert: state.alert,
    selectedVideo: getSelectedVideoWithFilters(state),
    isAuth: state.auth.isAuth
});

const mapDispatchToProps = (dispatch) => bindActionCreators({
    checkIsScanning,
    startFileScan,
    showErrorAlert,
    loadFilterOptions,
    hideAlert,
    saveVideoFileEdits,
    checkAuth,
    logout
}, dispatch);

const AppContentWrapped = connect(mapStateToProps, mapDispatchToProps)(AppContent);
AppContentWrapped.propTypes = {};
export default AppContentWrapped;