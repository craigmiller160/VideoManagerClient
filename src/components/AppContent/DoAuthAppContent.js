import React, { useEffect, useState } from 'react';
import { useDispatch } from 'react-redux';
import { checkAuth } from '../../store/auth/auth.actions';
import Spinner from '../UI/Spinner/Spinner';
import { AppContent } from './AppContent';

const DoAuthAppContent = () => {
    const dispatch = useDispatch();
    const [isChecking, setIsChecking] = useState(true);
    useEffect(() => {
        (async () => {
            await dispatch(checkAuth());
            setIsChecking(false);
        })();
    }, []);

    if (isChecking) {
        return <Spinner />
    }

    return <AppContent />;
};

export default DoAuthAppContent;