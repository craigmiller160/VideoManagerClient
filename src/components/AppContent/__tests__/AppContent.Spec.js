import React from 'react';
import { AppContent } from '../AppContent';
import { MemoryRouter, withRouter } from 'react-router-dom';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';

jest.mock('../../UI/Alert/Alert', () => () => 'Alert');
jest.mock('../VideoListLayout/VideoListLayout', () => () => 'VideoListLayout');

const createComponent = (props = {}, route = '/') => {
    const AppContentRouter = withRouter(AppContent);

    return mount(
        <MemoryRouter initalEntries={ [ route ] }>
            <AppContentRouter { ...props } />
        </MemoryRouter>
    );
};

const checkIsScanning = jest.fn();
const showErrorAlert = jest.fn(ex => console.log(ex));
const props = {
    checkIsScanning,
    showErrorAlert
};

describe('AppContent', () => {
    it('renders successfully with VideoListComponent for root route', (done) => {
        const component = createComponent(props);

        setImmediate(() => {
            expect(checkIsScanning).toHaveBeenCalled();
            expect(showErrorAlert).not.toHaveBeenCalled();
            expect(component.find('AppContent').state()).toEqual(expect.objectContaining({
                isStarted: true
            }));
            component.update();
            expect(component.find('AppContent Container').length).toEqual(2);
            expect(component.find('AppContent Route').text().trim()).toEqual('VideoListLayout');
            done();
        });
    });

    it('renders Scanning component for /scanning route', (done) => {
        const component = createComponent(props, '/scanning');

        setImmediate(() => {
            expect(checkIsScanning).toHaveBeenCalled();
            expect(showErrorAlert).not.toHaveBeenCalled();
            expect(component.find('AppContent').state()).toEqual(expect.objectContaining({
                isStarted: true
            }));
            component.update();
            expect(component.find('AppContent Container').length).toEqual(2);
            expect(component.find('AppContent Route').text().trim()).toEqual('Scanning');
            done();
        });
    });
});