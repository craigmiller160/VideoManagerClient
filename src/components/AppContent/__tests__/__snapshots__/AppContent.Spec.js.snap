// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppContent renders successfully with VideoListComponent for root route 1`] = `
<AppContent
  checkIsScanning={
    [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "isThrow": false,
          "value": undefined,
        },
      ],
    }
  }
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "ow91kq",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "ow91kq",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
  location={
    Object {
      "hash": "",
      "key": "ow91kq",
      "pathname": "/",
      "search": "",
      "state": undefined,
    }
  }
  match={
    Object {
      "isExact": true,
      "params": Object {},
      "path": "/",
      "url": "/",
    }
  }
  showErrorAlert={[MockFunction]}
>
  <div>
    <VideoNavbar>
      <Navbar
        color="dark"
        dark={true}
        expand="md"
        tag="nav"
      >
        <nav
          className="navbar navbar-expand-md navbar-dark bg-dark"
        >
          <Container
            tag="div"
          >
            <div
              className="container"
            >
              <NavbarBrand
                tag={[Function]}
                to="/"
              >
                <Link
                  className="navbar-brand"
                  replace={false}
                  to="/"
                >
                  <a
                    className="navbar-brand"
                    href="/"
                    onClick={[Function]}
                  >
                    Video Manager
                  </a>
                </Link>
              </NavbarBrand>
              <NavbarToggler
                onClick={[Function]}
                tag="button"
                type="button"
              >
                <button
                  className="navbar-toggler"
                  onClick={[Function]}
                  type="button"
                >
                  <span
                    className="navbar-toggler-icon"
                  />
                </button>
              </NavbarToggler>
              <Collapse
                appear={false}
                enter={true}
                exit={true}
                in={false}
                isOpen={false}
                mountOnEnter={false}
                navbar={true}
                onEnter={[Function]}
                onEntered={[Function]}
                onEntering={[Function]}
                onExit={[Function]}
                onExited={[Function]}
                onExiting={[Function]}
                tag="div"
                timeout={350}
                unmountOnExit={false}
              >
                <Transition
                  appear={false}
                  enter={true}
                  exit={true}
                  in={false}
                  mountOnEnter={false}
                  onEnter={[Function]}
                  onEntered={[Function]}
                  onEntering={[Function]}
                  onExit={[Function]}
                  onExited={[Function]}
                  onExiting={[Function]}
                  timeout={350}
                  unmountOnExit={false}
                >
                  <div
                    className="collapse navbar-collapse"
                    style={Object {}}
                  >
                    <Nav
                      className="ml-auto"
                      navbar={true}
                      tag="ul"
                      vertical={false}
                    >
                      <ul
                        className="ml-auto navbar-nav"
                      >
                        <NavItem
                          tag="li"
                        >
                          <li
                            className="nav-item"
                          >
                            <NavLink
                              id="scanDirectory"
                              onClick={[Function]}
                              tag="a"
                            >
                              <a
                                className="nav-link"
                                id="scanDirectory"
                                onClick={[Function]}
                              >
                                Scan Directory
                              </a>
                            </NavLink>
                          </li>
                        </NavItem>
                      </ul>
                    </Nav>
                  </div>
                </Transition>
              </Collapse>
            </div>
          </Container>
        </nav>
      </Navbar>
    </VideoNavbar>
    <Container
      tag="div"
    >
      <div
        className="container"
      >
        <Row
          tag="div"
        >
          <div
            className="row"
          >
            <Col
              tag="div"
              widths={
                Array [
                  "xs",
                  "sm",
                  "md",
                  "lg",
                  "xl",
                ]
              }
              xs={
                Object {
                  "offset": 2,
                  "size": 8,
                }
              }
            >
              <div
                className="col-8 offset-2"
              >
                <Component>
                  Alert
                </Component>
              </div>
            </Col>
          </div>
        </Row>
        <Switch>
          <Route
            computedMatch={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            exact={true}
            location={
              Object {
                "hash": "",
                "key": "ow91kq",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            path="/"
            render={[Function]}
          >
            <Component
              history={
                Object {
                  "action": "POP",
                  "block": [Function],
                  "canGo": [Function],
                  "createHref": [Function],
                  "entries": Array [
                    Object {
                      "hash": "",
                      "key": "ow91kq",
                      "pathname": "/",
                      "search": "",
                      "state": undefined,
                    },
                  ],
                  "go": [Function],
                  "goBack": [Function],
                  "goForward": [Function],
                  "index": 0,
                  "length": 1,
                  "listen": [Function],
                  "location": Object {
                    "hash": "",
                    "key": "ow91kq",
                    "pathname": "/",
                    "search": "",
                    "state": undefined,
                  },
                  "push": [Function],
                  "replace": [Function],
                }
              }
              location={
                Object {
                  "hash": "",
                  "key": "ow91kq",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                }
              }
              match={
                Object {
                  "isExact": true,
                  "params": Object {},
                  "path": "/",
                  "url": "/",
                }
              }
            >
              VideoListLayout
            </Component>
          </Route>
        </Switch>
      </div>
    </Container>
  </div>
</AppContent>
`;
